### PR TITLE
Fix a regression issue of parsing datetime string with custom time format in Span

### LINKED ISSUE #3078: [BUG] PPL gets error but looks syntax correct
**What is the bug?**
I run a syntax correct PPL but it shows a bug. It works on 2.15 but fails on 2.17 and current main

**How can one reproduce the bug?**
Index mappings: 
```
{
"mappings": {
      "properties": {
        "Content": {
          "type": "text"
        },
        "Level": {
          "type": "keyword"
        },
        "OriginalLog": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword"
            }
          }
        },
        "Time": {
          "type": "date",
          "format": "EEE MMM dd HH:mm:ss yyyy||yyyy-MM-dd||epoch_millis"
        }
      }
    }
}
```
PPL: 
`"source=loghub-apache-new | where Level = 'error' | eval is_mod = IF(LIKE(OriginalLog, '%mod_jk%'), 1, 0), is_dir = IF(LIKE(OriginalLog, '%directory%'), 1, 0) | stats COUNT() as sum_is_mod by span(Time, 1d)"`

Some samples:
```
{
          "Time": "Sun Dec 04 06:19:18 2005",
          "Level": "error",
          "Content": "mod_jk child workerEnv in error state 6",
          "OriginalLog": "[Sun Dec 04 06:19:18 2005] [error] mod_jk child workerEnv in error state 6"
        }
        {
          "Time": "Sun Dec 04 06:22:48 2005",
          "Level": "notice",
          "Content": "jk2_init() Found child 32479 in scoreboard slot 6",
          "OriginalLog": "[Sun Dec 04 06:22:48 2005] [notice] jk2_init() Found child 32479 in scoreboard slot 6"
        }
```

**What is the expected behavior?**

**What is your host/environment?**
OS 2.17

**Do you have any screenshots?**

**Do you have any additional context?**
Add any other context about the problem.

### PR DESCRIPTION
**Pull request by LantaoJin**
### Description
This seems a regression issue from https://github.com/opensearch-project/sql/pull/2762. Query of StatsBy with Span on a custom formatted date type will fail.

Here is the reproduce:
```
PUT /ltjin-test1
{
    "mappings":{
        "properties": {
            "Time1": {
                "type": "date",
                "format": "yyyy-MM-dd"
            },
            "Time2": {
                "type": "date",
                "format": "epoch_millis"
            }
        }
    }
}

POST _bulk
{ "index" : { "_index" : "ltjin-test1", "_id" : "1" } }
{ "Time1": "2005-12-05", "Time2": 1133654400000 }

POST /_plugins/_ppl/
{
    "query": """
    source=ltjin-test1 | eval a = 1 | stats Count(a) as a by span(Time1, 1d)
    """
}

POST /_plugins/_ppl/
{
    "query": """
    source=ltjin-test1 | eval a = 1 | stats Count(a) as a by span(Time2, 1d)
    """
}
```
Throws NPE.

### Related Issues
Resolves #3078 

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
 - [ ] New functionality has javadoc added.
 - [ ] New functionality has a user manual doc added.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by YANG-DB — DISMISSED** [review]

**Review by MaxKsyunz — COMMENTED** [review]
Please increase unit test coverage and resolve now failing integration tests.

```
-----------
* What went wrong:
Execution failed for task ':core:jacocoTestCoverageVerification'.
> Rule violated for class org.opensearch.sql.planner.physical.collector.Rounding: branches covered ratio is 0.8, but expected minimum is 1.0
```

and make several integration tests fail
```
DateTimeFormatsIT > testIncompleteFormats FAILED
    java.lang.AssertionError: 
    Expected: iterable with items [[1984-01-01 00:00:00, null, null, 10:00:00, 1999-01-01], [2012-01-01 00:00:00, null, null, 20:00:00, 3021-01-01]] in any order
         but: not matched: <["1984-01-01 00:00:00",null,null,"1970-01-01","1999-01-01"]>
        at __randomizedtesting.SeedInfo.seed([2C68C17CE23C9F5D:B917EF6B60642A6C]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.opensearch.sql.util.MatcherUtils.verify(MatcherUtils.java:188)
        at org.opensearch.sql.util.MatcherUtils.verifyDataRows(MatcherUtils.java:159)
        at org.opensearch.sql.sql.DateTimeFormatsIT.testIncompleteFormats(DateTimeFormatsIT.java:124)

DateTimeFormatsIT > testCustomFormats FAILED
    java.lang.AssertionError: 
    Expected: iterable with items [[09:07:42, 1984-04-12 09:07:42, 1984-04-12, 1961-04-12 00:00:00, 23:44:36.321], [21:07:42, 1984-04-12 22:07:42, 1984-04-12, 1970-01-01 09:07:00, 09:01:16.542]] in any order
         but: not matched: <["09:07:42","1984-04-12 09:07:42","1984-04-12","1961-04-12","1970-01-01"]>
        at __randomizedtesting.SeedInfo.seed([2C68C17CE23C9F5D:203B87710DB1C56E]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.opensearch.sql.util.MatcherUtils.verify(MatcherUtils.java:188)
        at org.opensearch.sql.util.MatcherUtils.verifyDataRows(MatcherUtils.java:159)
        at org.opensearch.sql.sql.DateTimeFormatsIT.testCustomFormats(DateTimeFormatsIT.java:76)

DateTimeFormatsIT > testDateFormatsWithOr FAILED
    java.lang.AssertionError: 
    Expected: iterable with items [[1984-04-12 00:00:00], [1984-04-12 09:07:42.000123456]] in any order
         but: not matched: <["1984-04-12"]>
        at __randomizedtesting.SeedInfo.seed([2C68C17CE23C9F5D:51D930CA323C2C64]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.opensearch.sql.util.MatcherUtils.verify(MatcherUtils.java:188)
        at org.opensearch.sql.util.MatcherUtils.verifyDataRows(MatcherUtils.java:159)
        at org.opensearch.sql.sql.DateTimeFormatsIT.testDateFormatsWithOr(DateTimeFormatsIT.java:57)

DateTimeFormatsIT > testDateNanosWithNanos PASSED

DateTimeFormatsIT > testCustomFormats2 FAILED
    java.lang.AssertionError: 
    Expected: iterable with items [[1984-10-20, 10:20:30, 1984-10-20 15:35:48], [1961-04-12, 09:07:00, 1961-04-12 09:07:00]] in any order
         but: not matched: <["1970-01-01","1970-01-01","2598-09-26"]>
        at __randomizedtesting.SeedInfo.seed([2C68C17CE23C9F5D:75DB39273E844970]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.opensearch.sql.util.MatcherUtils.verify(MatcherUtils.java:188)
        at org.opensearch.sql.util.MatcherUtils.verifyDataRows(MatcherUtils.java:159)
        at org.opensearch.sql.sql.DateTimeFormatsIT.testCustomFormats2(DateTimeFormatsIT.java:102)
```

**Comment by LantaoJin** [comment]
> Please increase unit test coverage and resolve now failing integration tests.

The unit test coverage can't be fixed since jacoco report is module-level based. What's the bug here is missing the `OpenSearchDateType` handling in `Rounding` logic , but `OpenSearchDateType` is not in the core module.
I fixed the failures in `DateTimeFormatsIT` and add a case to this IT.

**Comment by LantaoJin** [comment]
@YANG-DB @MaxKsyunz @penghuo , since this seems a regression issue introduced from v2.16.0. I'd like this fixing could be merged in v2.18.0

File: core/src/main/java/org/opensearch/sql/planner/physical/collector/Rounding.java [hunk 1]
```diff
@@ -46,13 +46,13 @@ public static Rounding<?> createRounding(SpanExpression span) {
     if (DOUBLE.isCompatible(type)) {
       return new DoubleRounding(interval);
     }
-    if (type.equals(TIMESTAMP)) {
+    if (type.equals(TIMESTAMP) || type.typeName().equalsIgnoreCase(TIMESTAMP.typeName())) {
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/physical/collector/Rounding.java:49]: what case covered in here?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java [hunk 2]
```diff
@@ -295,7 +295,7 @@ private static ExprValue createOpenSearchDateType(Content value, ExprType type)
         }
       } else {
         // custom format
-        return parseDateTimeString(value.stringValue(), dt);
+        return parseDateTimeString(value.objectValue().toString(), dt);
```
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java:298]: could u elberate more? value is number value? why objectValue().toString()?

**Comment by penghuo** [comment]
could u exclude Rounding from TestCoverage and address it later. otherwise, following PR will failed with test coverage?

File: core/src/main/java/org/opensearch/sql/planner/physical/collector/Rounding.java (hunk 1 above)
Comment (LantaoJin) [reply core/src/main/java/org/opensearch/sql/planner/physical/collector/Rounding.java:49]: `type` could be an `OpenSearchDateType`, rather `ExprCoreType` after #2762

File: opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java (hunk 2 above)
Comment (LantaoJin) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java:298]: The value is an instance of `ObjectContent` and isNumber() is true, `stringValue()` will return the string representation of the object, such as `ObjectContent@1234`. We should convert it to Java Object then use `toString()` here. Or `value.longValue().toString()`. Not sure the value here is a Long type, so `value.objectValue().toString()` would be best.

**Comment by LantaoJin** [comment]
> could u exclude Rounding from TestCoverage and address it later. otherwise, following PR will failed with test coverage?

Excluding `Rounding` is not what we expected, there are many logic in class `Rounding`, IMO jacoco only fail the incremental codes which is less than the threshold of coverage, the following PR shouldn't be impacted.

**Review by YANG-DB — DISMISSED** [review]

**Comment by LantaoJin** [comment]
I added some unit tests in a mock way to fix the missing test coverage. Now CI passed! Please take a review again @YANG-DB @MaxKsyunz @penghuo @dai-chen @ykmr1224. This patch could be marked as critical for 2.18.0!

**Review by penghuo — APPROVED** [review]
Thx!

**Review by seankao-az — APPROVED** [review]

**Pull request merged**